### DISCUSSION TITLE
How to update footprint from library in prefight?

### DISCUSSION THREAD

**Discussion by dariuskt** [INTI-CMNB/KiBot#483] [Q&A] [answered]
I generate qr code externally and convert it to footprint before running KiBot. But then I run KiBot I get the old footprint in all the renders. Is there a way to refresh a particular footprint before every run?

**Comment by set-soft** [comment]
Hi @dariuskt !
The QR is a common case.
Why don't you use the QR facilities offered by KiBot? Using the `update_qr` preflight, that needs a `qr_lib` output, you'll get it.
The way that KiCad currently works updating a footprint (and/or symbol) implies a modification to the PCB (and/or schematic).
KiBot takes care of this in the `qr_lib` output, that can be promoted to a preflight using the `update_qr` preflight.

**Comment by dariuskt** [comment]
Well, I tried qr_lib but had a few problems with it. at first, I couldn't get it to make just a single kicad_mod file without directories. giving it output as a filename made a file that kicad6 complained about syntax errors, and after all this enabling update_qr just crashed KiBot every time with a Python error.  At this point, I thought there surely must be a better way to update a footprint from the library as QR generating outside KiBot gives me more flexibility (I embed my logo in the QR code).

As for modifying pcb, I think fill_zones does that as well, so I see no problem having the option to update a footprint. Maybe footprint update ability could be exposed somehow independently from the qr_lib? This topic kinda unintentionally turned into a feature request ;)

As you didn't mention a way to update the footprint independently I assume there is no way to do that.

**Comment by set-soft** [comment]
> Well, I tried qr_lib but had a few problems with it. 

If you don't report problems they won't get fixed ...

> at first, I couldn't get it to make just a single kicad_mod file without directories. giving it output as a filename made a file that kicad6 complained about syntax errors, and after all this enabling update_qr just crashed KiBot every time with a Python error. 

The qr_lib output is covered by tests, so it works for at least one case. If you have a case where it fails just report it.

> At this point, I thought there surely must be a better way to update a footprint from the library as QR generating outside KiBot gives me more flexibility (I embed my logo in the QR code).

Don't you think other people could also need it? KiBot grows from suggestions to solve problems for projects.

> As for modifying pcb, I think fill_zones does that as well, so I see no problem having the option to update a footprint. Maybe footprint update ability could be exposed somehow independently from the qr_lib? This topic kinda unintentionally turned into a feature request ;)

Yes. But I think fixing/extending qr_lib is the way to go.

> As you didn't mention a way to update the footprint independently I assume there is no way to do that.

You are right.

**Comment by dariuskt** [comment]
About reporting problems, I'm fairly new to KiCad and all the plugins and KiBot, so I don't have a reasonable understanding of which part belongs to what. So I have no idea is that integration problem or is it upstream lib problem or most likely my configuration problem. If you care to see I could prepare a demo project with easily reproducible problem I experienced.

About extending qr_lib, that definitely would be great, but at some point, qr_lib might be not the right name :D my current use case could be addressed by:
* allowing to update footprint without generating anything - most flexible feature for all different use cases we can't even imagine today
* generating footprint from user provided image instead of text for QR code - that would also be fairly flexible as the user can supply custom graphics, maybe barcode or whatever ;)
* Adding the ability to embed logo into the center of QR code from user-provided image - that would solve my current case but not really flexible beond that

**Comment by set-soft** [comment]
> About reporting problems, I'm fairly new to KiCad and all the plugins and KiBot, so I don't have a reasonable understanding of which part belongs to what. So I have no idea is that integration problem or is it upstream lib problem or most likely my configuration problem. If you care to see I could prepare a demo project with easily reproducible problem I experienced.

Yes, please create a project. If you can do it in a repo better.

**Comment by dariuskt** [comment]
I made a repo to better illustrate what I'm talking about. I include my naive attempts which might be rectified in the documentation or some higher-level error messages.

https://github.com/dariuskt/tmp-qr_lib/tree/minimal_config
https://github.com/dariuskt/tmp-qr_lib/tree/output_path

looks like all my problems arise from attempts to generate only footprint without symbol and place that footprint in the directory I want without any extra directories.

looking forward to your thoughts on this.

**Reply by set-soft** [reply]
I'll be busy some days, but will take a look at this in a couple of weeks, please be patient

**Reply by set-soft** [reply]
> I made a repo to better illustrate what I'm talking about. I include my naive attempts which might be rectified in the documentation or some higher-level error messages.
> 
> https://github.com/dariuskt/tmp-qr_lib/tree/minimal_config 
> https://github.com/dariuskt/tmp-qr_lib/tree/output_path

> 
> looks like all my problems arise from attempts to generate only footprint without symbol and place that footprint in the directory I want without any extra directories.

Yes, this is a real mess. You are forcing a lot of things to go really wrong.

1. This output will generate a *symbol* and a *footprint*, not just a footprint. This is clearly documented.
2. You **must** use the _%x_ var somewhere in the name. Now KiBot will just stop if you don't do it. In this way the symbol lib and the footprint lib won't collide.
3. KiCad concept of a footprint lib isn't a single file floating around, like it was in KiCad 4. A library is a directory, named *xxxx.pretty* containing one or more footprints (*xxxx.kicad_mod*). This output creates libraries, you can specify one or more QRs. So you are asking to create a lib with wrong name overwriting an existing file, which generates an IO error.

I added more clarifications to the docs:

```
Filename/dirname for the output library (%i=qr, %x=lib/kicad_sym/pretty).
You must use %x in the name to get a symbols lib and a footprints lib
 ```

**Comment by dariuskt** [comment]
I'm glad we achieved improvement ;)

1. Most of the time if documentation says that tool can produce two independent things - it also can produce any of the things by itself... so maybe that part could be more clear that making only a footprint or symbol is not possible.
2. good, this behavior would have saved me a lot of time.
3. It wasn't clear to me that only a library(footprint in a new directory) can be created and a footprint in an existing library specified by a user can not. 
> This output creates a library containing a symbol and footprint for a QR code.

So with this tool and configuration being so rigid, it would be pretty hard to use in my case. 
* I would need to run KiBot for only qr_lib output
* then remove the symbol as I don't need it
* move footprint to my original footprints dir
* remove generated empty dir
* run KiBot normally just without qr_lib output

Instead I just
* create QR with `qrencode`
* add logo with `imagemagick`
* convert to footprint with `img2mod`
* run KiBot normally

In my setup looks like only `img2mod` really wants to be managed by KiBot as long as it would not gain any extra limitations.
Hope my use case is clear and you can come to an informed decision whether this should be compatible with KiBot or not.

**Comment by set-soft** [comment]
> So with this tool and configuration being so rigid, it would be pretty hard to use in my case.
> 
> * I would need to run KiBot for only qr_lib output

Why? 
Yes, you need to run KiBot, not necessarily to just generate the QR. And you need to do it at least once.
Depending on the QR content you'll have to run KiBot often: yes, to update it.
So I don't see any issue here.

> * then remove the symbol as I don't need it

Why?
This is very bad practice, every footprint should be associated to a symbol. This makes things much more simple and avoids confusion.
Even if you choose not to use it, why remove it? is a very small file.

> * move footprint to my original footprints dir
> * remove generated empty dir

Why?
You can just add the KiBot lib to your *fp-lib-table*/*symbol-lib-table*. This is just another lib. And is good to have it separated because this is generated.

> * run KiBot normally just without qr_lib output

Why?
The qr_lib can be run as a pre-flight. You don't need to run things separately.

Again: you are forcing a workflow that is too complicated.


> Instead I just
> 
> * create QR with `qrencode`
> * add logo with `imagemagick`
> * convert to footprint with `img2mod`
> * run KiBot normally

I think this could be just part of the current qr_lib output.

> In my setup looks like only `img2mod` really wants to be managed by KiBot as long as it would not gain any extra limitations. Hope my use case is clear and you can come to an informed decision whether this should be compatible with KiBot or not.

1. I think that adding an option to update a footprint is a good idea. This is why I added #492 as a feature request.
2. I also think that what you really need is more functionality in qr_lib and look a little bit closer to how to use it. 
 
You don't need separated runs, you don't need to move files and you don't need to remove files. Just add the preflight and add the generated libs to your project. And yes, you'll need to run KiBot at least once to get the first QR.

**Comment by set-soft** [comment, marked as answer]
Now we have `update_footprint` preflight.
